src/test/java/net/openhft/chronicle/core/io/ReferenceCountedContractTest.java

```diff
@@ -0,0 +1,242 @@
+        referenceCounted.reserve(b);
+        assertEquals(3, referenceCounted.refCount());
+
+        referenceCounted.release(b);
+        assertEquals(2, referenceCounted.refCount());
+
+        referenceCounted.release(a);
+        assertEquals(1, referenceCounted.refCount());
+    }
+
+    @Test
+    public void releaseWillFailWhenResourceAlreadyReleased() {
+        ReferenceCounted referenceCounted = createReferenceCounted();
+
+        referenceCounted.releaseLast();
+
+        ReferenceOwner a = ReferenceOwner.temporary("a");
+        assertThrows(IllegalStateException.class, () -> referenceCounted.release(a));
+    }
+
+    @Test
+    public void releaseWillGoAllTheWayToZero() {
+        ReferenceCounted referenceCounted = createReferenceCounted();
+
+        referenceCounted.release(ReferenceOwner.INIT);
+        assertEquals(0, referenceCounted.refCount());
+    }
+
+    @Test
+    public void releaseLastWillDecrementReferenceCount() {
+        ReferenceCounted referenceCounted = createReferenceCounted();
+
+        assertEquals(1, referenceCounted.refCount());
+
+        referenceCounted.releaseLast();
+        assertEquals(0, referenceCounted.refCount());
+    }
+
+    @Test
+    public void releaseLastWillReleaseThenFailWhenReferenceIsNotLast() {
+        ReferenceCounted referenceCounted = createReferenceCounted();
+
+        ReferenceOwner a = ReferenceOwner.temporary("a");
+        referenceCounted.reserve(a);
+
+        assertEquals(2, referenceCounted.refCount());
+
+        assertThrows(IllegalStateException.class, referenceCounted::releaseLast);
+
+        assertEquals(1, referenceCounted.refCount());
+    }
+
+    @Test
+    public void releaseLastWillFailWhenResourceAlreadyReleased() {
+        ReferenceCounted referenceCounted = createReferenceCounted();
+
+        referenceCounted.releaseLast();
+        assertThrows(IllegalStateException.class, referenceCounted::releaseLast);
+    }
+
+    @Test
+    public void tryReserveWillReturnTrueWhenReservationWasSuccessful() {
+        ReferenceCounted referenceCounted = createReferenceCounted();
+
+        ReferenceOwner a = ReferenceOwner.temporary("a");
+        assertTrue(referenceCounted.tryReserve(a));
+    }
+
+    @Test
+    public void tryReserveWillReturnFalseWhenResourceIsAlreadyReleased() {
+        ReferenceCounted referenceCounted = createReferenceCounted();
+
+        referenceCounted.releaseLast();
+        ReferenceOwner a = ReferenceOwner.temporary("a");
+        assertFalse(referenceCounted.tryReserve(a));
+    }
+
+    @Test
+    public void reservedByWillReturnTrueWhenOwnerHasReferenceReserved() {
+        ReferenceCounted referenceCounted = createReferenceCounted();
+
+        ReferenceOwner a = ReferenceOwner.temporary("a");
+        referenceCounted.reserve(a);
+        assertTrue(referenceCounted.reservedBy(a));
+    }
+
+    @Test
+    public void implementationsShouldBeThreadSafe() throws InterruptedException {
```

Comment:
Added this to weed out any thread-safety issues in ReferenceCounted implementations